hal/device.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef HAL_DEVICE_H
+#define HAL_DEVICE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include "defined_error.h"
+
+/**
+ * Device abstract layer
+ * Implement your device inside devices
+ */
+typedef struct device_s device_t;
+struct device_s {
+  void (*destructor)(device_t *);
```

Comment:
Shall we provide `init`?

Reply:
How about changing from `destructor` to `fini`?

Reply:
> Shall we provide `init`?

Will it be helpful?  How will the object be generated after we improve the `init` inside device_s class?

Reply:
If we provide `init`, we can eliminate the assignment for UART device (such as `ttyUSB0`) where we can focus on the logic behind device manipulation.